### PR TITLE
Enable local disk config reload

### PR DESCRIPTION
**Pull request by taiyang-li**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Enable reloading of local disk config.

Reproduce steps:
1. Update `keep_free_space_bytes` of disk `ebs_gp2_1` from `3072` to `1024` in clickhouse config file, then execute
for example: 
```
<disks>
      <default>
         <keep_free_space_bytes>1024</keep_free_space_bytes>
      </default>

      <ebs_gp2_1>
         <path>/data/clickhouse_2/</path>
         <keep_free_space_bytes>3072</keep_free_space_bytes>
      </ebs_gp2_1>

</disks>
```

2. Execute `SYSTEM CONFIG RELOAD`

3. Query disk config: `SELECT　name,　path,　formatReadableSize(free_space) AS free,　formatReadableSize(total_space) AS total,　formatReadableSize(keep_free_space) AS reserved　FROM system.disks`

4. `reserved` of local disk `ebs_gp2_1` is still `3072`
```
SELECT
    name,
    path,
    formatReadableSize(free_space) AS free,
    formatReadableSize(total_space) AS total,
    formatReadableSize(keep_free_space) AS reserved
FROM system.disks

Query id: 38a47b14-aacc-4578-a6bf-f1732ab5e661

┌─name──────┬─path──────────────────┬─free──────┬─total──────┬─reserved─┐
│ default   │ /data1/clickhouse_ly/ │ 62.12 TiB │ 72.47 TiB  │ 1.00 KiB │
│ ebs_gp2_1 │ /data/clickhouse_2/   │ 99.01 GiB │ 222.16 GiB │ 3.00 KiB │
└───────────┴───────────────────────┴───────────┴────────────┴──────────┘
```

### CODE REVIEW

File: src/Disks/DiskSelector.cpp [hunk 1]
```diff
@@ -77,6 +77,9 @@ DiskSelectorPtr DiskSelector::updateFromConfig(
 
             /// TODO: Ideally ClickHouse shall complain if disk has changed, but
             /// implementing that may appear as not trivial task.
+            auto disk = std::static_pointer_cast<DiskLocal>(result->getDisksMap().find(disk_name)->second);
+            if (disk)
+                disk->updateFromConfig(config, disk_config_prefix, context);
```
Comment (azat) [inline src/Disks/DiskSelector.cpp:82]: Maybe this should be done in background?
Like other config directives reloading - https://github.com/ClickHouse/ClickHouse/blob/ccb7e6e03f75f243c956e2718f615fab344c0e64/programs/server/Server.cpp#L685

Or, maybe config should include some version, so that you can check against it, and do nothing if version is the same.

Also it should note that it will not work if DiskPtr is cached, and this includes lot's of engines (Distriubted, TInyLog, Log, ....), but AFAICS this will work for MergeTree engines.

Comment (taiyang-li) [reply src/Disks/DiskSelector.cpp:82]: 1. I think this reload is already running  in background. Invoking chain like below: 
```
main_config_reloader
-> Context::updateStorageConfiguration
-> DiskSelector::updateFromConfig
-> DiskLocal::updateFromConfig
```

2. Decide whether to reload by version is a good idea.  I will finish this optimization. 
3. It is `DiskPtr` that is referenced by engines. If the object which `DiskPtr` pointed to just updated from config, the updating is visible for engines.

Comment (azat) [reply src/Disks/DiskSelector.cpp:82]: >I think this reload is already running in background. Invoking chain like below:

Yep, but I though about updating it just from there, but forget it, config version will be enough

> It is DiskPtr that is referenced by engines. If the object which DiskPtr pointed to just updated from config, the updating is visible for engines.

You are right (I though that your patch creates new Disk)

Also have you considered not only disks but also volumes?

Comment (taiyang-li) [reply src/Disks/DiskSelector.cpp:82]: I have not thought about volumns yet. 

In the first step, I planed to enable disk reload(not only `DiskLocal`, but also `DiskMemory` `DiskS3` and so on).  
Then considered reloading of volumns.  
This looks like a huge plan, I will try to finish it in this weekend.

File: src/Disks/DiskLocal.cpp [hunk 2]
```diff
@@ -337,6 +391,22 @@ void DiskLocal::sync(int fd) const
...
+                        const Context & context)
+{
+    String new_disk_path;
+    UInt64 new_keep_free_space_bytes;
+    loadDiskLocalConfig(name, config, config_prefix, context, new_disk_path, new_keep_free_space_bytes);
+    if (disk_path != new_disk_path)
+        throw Exception("Disk path can't update from config " + name, ErrorCodes::UNKNOWN_ELEMENT_IN_CONFIG);
+
+    // Config of DiskLocal has been updated
+    if (keep_free_space_bytes != new_keep_free_space_bytes)
```
Comment (taiyang-li) [inline src/Disks/DiskLocal.cpp:405]: @azat  The only parameter in `DiskLocal` is `keep_free_space_bytes`,  maybe the number of parameters in `DiskLocal` is too few to juding whether to reload config by version?

Comment (azat) [reply src/Disks/DiskLocal.cpp:405]: @taiyang-li I was talking about more generic approach for config versioning:
- config should have some global version
- DiskLocal should cache it and call updateFromConfig only if the version of the config is newer then the cached version in DiskLocal

File: src/Disks/DiskSelector.cpp (hunk 1 above)
Comment (azat) [reply src/Disks/DiskSelector.cpp:82]: >It is DiskPtr that is referenced by engines. If the object which DiskPtr pointed to just updated from config, the updating is visible for engines.

Also note that if such storage (Log or similar) uses some unique volume, for which none will call updateFromConfig it will not be updated, that's why I was suggesting to update them in background (so that you explicitly iterate over all existing disks to update them all periodically, and in this case you may not need config version, since this will add overhead only into this background update)

File: src/Disks/DiskLocal.cpp (hunk 2 above)
Comment (taiyang-li) [reply src/Disks/DiskLocal.cpp:405]: Got it.  I will update this code.

Comment (azat) [reply src/Disks/DiskLocal.cpp:405]: @taiyang-li  Also please take a look at https://github.com/ClickHouse/ClickHouse/pull/19526#discussion_r564302583 maybe you will prefer this way

File: src/Disks/DiskSelector.cpp (hunk 1 above)
Comment (taiyang-li) [reply src/Disks/DiskSelector.cpp:82]: > Also note that if such storage (Log or similar) uses some unique volume, for which none will call updateFromConfig it will not be updated, that's why I was suggesting to update them in background (so that you explicitly iterate over all existing disks to update them all periodically, and in this case you may not need config version, since this will add overhead only into this background update)

Sorry, besides no version compare,  I still didn't  understand the advantage of background schedule over another solution. Disk reload(at least for `DiskLocal`) is not a heavy operation, and I think putting it into `main_config_reloader` thread is not a big problem...

Comment (taiyang-li) [reply src/Disks/DiskSelector.cpp:82]: Since I didn't fully understand the points of  background schedule  Maybe I should try version solution firstly...

Comment (azat) [reply src/Disks/DiskSelector.cpp]: > I still didn't understand the advantage of background schedule over another solution.

Let's consider the following example:

- One external disk:

```xml
<storage_configuration>
  <disks>
    <external_for_log_data_table>
      <path>/tmp/</path>
      <keep_free_space_bytes>1000</keep_free_space_bytes>
    </external_for_log_data_table>
  </disks>
</storage_configuration>
```

- Log table, that uses this disk:

```sql
create table data (key Int) engine=Log() settings disk='external_for_log_data_table'
```

And this will create StorageLog object, with cached Disk, that will be never updated, until someone will get it via DiskSelector (or similar API).

But if you will update disks in background (i.e. call `DiskLocal::updateFromConfigIfChanged` for each existing disk somehow), then you don't have such problem, and the changes will be applied to such cached Disks too (for whom nobody call getDisk again).

Sure you can workaround the problem by detach/attach, or create another table with the same disk name to call update, but this does not looks sane

File: src/Disks/DiskLocal.cpp [hunk 3]
```diff
@@ -337,6 +391,22 @@ void DiskLocal::sync(int fd) const
...
+{
+    String new_disk_path;
+    UInt64 new_keep_free_space_bytes;
+    loadDiskLocalConfig(name, config, config_prefix, context, new_disk_path, new_keep_free_space_bytes);
+
+    if (disk_path != new_disk_path)
+        throw Exception("Disk path can't update from config " + name, ErrorCodes::UNKNOWN_ELEMENT_IN_CONFIG);
+        
+    if (keep_free_space_bytes != new_keep_free_space_bytes)
+        keep_free_space_bytes = new_keep_free_space_bytes;
```
Comment (azat) [inline src/Disks/DiskLocal.cpp:406]: Also AFAICS this is racy, am I right?

Comment (taiyang-li) [reply src/Disks/DiskLocal.cpp:406]: Yes, you are right.  Maybe `keep_free_space_bytes` with atomic type could avoid racy.

Comment (taiyang-li) [reply src/Disks/DiskLocal.cpp:406]: done

File: src/Disks/DiskSelector.cpp (hunk 1 above)
Comment (taiyang-li) [reply src/Disks/DiskSelector.cpp]: I think StorageLog will get the updated disk after reloading disk config. See code: 
```
registerStorageLog -> Context::getDisk -> Context::getDiskSelector -> DiskSelector::get -> StorageLog::create  
```
`DiskSelector` contains all the disks defined in config files, including the disk used by `StorageLog`.  In this PR, when I execute `DiskSelector::updateFromConfig`, all the shard disk objects are updated if changed.

Comment (azat) [reply src/Disks/DiskSelector.cpp]: @taiyang-li Yes, you are right (I though about different call chain by some reason).

BTW looks like it worth adding some test for this.

File: src/Disks/DiskSelector.cpp [hunk 4]
```diff
@@ -77,6 +77,9 @@ DiskSelectorPtr DiskSelector::updateFromConfig(
 
             /// TODO: Ideally ClickHouse shall complain if disk has changed, but
             /// implementing that may appear as not trivial task.
+
+            /// TODO: Implete updateFromConfigIfChanged for DiskS3
```
Comment (azat) [inline src/Disks/DiskSelector.cpp:81]:
```suggestion
            /// TODO: Implement updateFromConfigIfChanged for DiskS3
```

File: src/Disks/DiskSelector.cpp [hunk 5]
```diff
@@ -77,6 +77,9 @@ DiskSelectorPtr DiskSelector::updateFromConfig(
 
             /// TODO: Ideally ClickHouse shall complain if disk has changed, but
             /// implementing that may appear as not trivial task.
```
Comment (azat) [inline src/Disks/DiskSelector.cpp:79]: This TODO looks irrelevant already (or it should be rewritten in another formulation)

P.S. sorry to bother, but you have conflicts with upstream, so you will need to fix the first anyway and this comment can be adjusted together

File: src/Disks/DiskLocal.cpp (hunk 3 above)
Comment (azat) [reply src/Disks/DiskLocal.cpp:406]: @taiyang-li btw can you resolve comments that had been already resolved?

File: src/Disks/DiskSelector.cpp (hunk 5 above)
Comment (taiyang-li) [reply src/Disks/DiskSelector.cpp]: Already fixed

File: src/Disks/DiskSelector.cpp (hunk 4 above)
Comment (taiyang-li) [reply src/Disks/DiskSelector.cpp]: Done.

File: src/Disks/DiskSelector.cpp (hunk 1 above)
Comment (taiyang-li) [reply src/Disks/DiskSelector.cpp]: Test already added.

**Comment by vdimir** [comment]
Similar for S3 disks https://github.com/ClickHouse/ClickHouse/pull/23429

**Review by kssenii — APPROVED** [review]
